silence/silence.go

```diff
@@ -611,43 +602,55 @@ func (s *Silences) Set(sil *pb.Silence) error {
 	defer s.mtx.Unlock()
 
 	now := s.nowUTC()
-	prev, ok := s.getSilence(sil.Id)
+	if sil.StartsAt.IsZero() {
+		sil.StartsAt = now
+	}
+
+	if err := validateSilence(sil); err != nil {
+		return fmt.Errorf("invalid silence: %w", err)
+	}
+
+	existing, ok := s.getSilence(sil.Id)
 	if sil.Id != "" && !ok {
 		return ErrNotFound
 	}
 
-	if ok && canUpdate(prev, sil, now) {
-		return s.setSilence(sil, now, false)
+	// updateExisting is true iff this is an existing silence, and the update does
+	// not affect how alerts are matched or the time window in which the
+	// silence is active (excluding time extensions).
+	updateExisting := ok && canUpdate(existing, sil, now)
+	if !updateExisting {
+		// If this is a new silence, or it is replacing an existing silence,
+		// then give it an ID and make sure the start time is not in the past.
+		uid, err := uuid.NewV4()
+		if err != nil {
+			return fmt.Errorf("failed to create UUID: %w", err)
+		}
+		sil.Id = uid.String()
+		if sil.StartsAt.Before(now) {
+			sil.StartsAt = now
+		}
 	}
+	sil.UpdatedAt = now
 
-	// If we got here it's either a new silence or a replacing one (which would
-	// also create a new silence) so we need to make sure we have capacity for
-	// the new silence.
-	if s.limits.MaxSilences != nil {
-		if m := s.limits.MaxSilences(); m > 0 && len(s.st)+1 > m {
-			return fmt.Errorf("exceeded maximum number of silences: %d (limit: %d)", len(s.st), m)
-		}
+	// Check the limits before expiring the existing silence to prevent
+	// partial updates when limits are exceeded. Limits are checked against
+	// the mesh silence as this is the silence that is gossiped and stored
+	// on disk.
+	msil := s.toMeshSilence(sil)
+	if err := s.checkLimits(msil, updateExisting); err != nil {
+		return err
 	}
 
-	if ok && getState(prev, s.nowUTC()) != types.SilenceStateExpired {
+	if ok && !updateExisting && getState(existing, s.nowUTC()) != types.SilenceStateExpired {
```

Comment:
As we mentioned this morning, I don't think we need the check for `getState(existing, s.nowUTC()) != types.SilenceStateExpired`. But I don't want to introduce a possible regression here, so I will look into that another time.